src/pybind/mgr/dashboard/controllers/rgw.py

```diff
@@ -84,40 +84,182 @@ def get(self, svc_id):
+            raise DashboardException(e, http_status_code=500, component='rgw')
 
 
 @ApiController('/rgw/bucket')
 @AuthRequired()
-class RgwBucket(RESTController):
+class RgwBucket(RgwProxy):
+
+    def list(self):
+        return self.proxy('GET', 'bucket')
+
+    def get(self, bucket):
+        return self.proxy('GET', 'bucket', {'bucket': bucket})
 
     def create(self, bucket, uid):
         try:
             rgw_client = RgwClient.instance(uid)
             return rgw_client.create_bucket(bucket)
         except RequestException as e:
-            cherrypy.response.headers['Content-Type'] = 'application/json'
-            cherrypy.response.status = 500
-            return {'detail': str(e)}
+            raise DashboardException(e, http_status_code=500, component='rgw')
+
+    def set(self, bucket, bucket_id, uid):
+        return self.proxy('PUT', 'bucket', {
+            'bucket': bucket,
+            'bucket-id': bucket_id,
+            'uid': uid
+        }, json_response=False)
+
+    def delete(self, bucket, purge_objects='true'):
+        return self.proxy('DELETE', 'bucket', {
+            'bucket': bucket,
+            'purge-objects': purge_objects
+        }, json_response=False)
+
+
+@ApiController('/rgw/user')
+@AuthRequired()
+class RgwUser(RgwProxy):
+
+    def list(self):
+        return self.proxy('GET', 'metadata/user')
+
+    def get(self, uid):
+        return self.proxy('GET', 'user', {'uid': uid})
+
+    def create(self, uid, display_name, email=None, max_buckets=None,
+               suspended=None, generate_key=None, access_key=None,
+               secret_key=None):
+        params = {'uid': uid}
+        if display_name is not None:
+            params['display-name'] = display_name
+        if email is not None:
+            params['email'] = email
+        if max_buckets is not None:
+            params['max-buckets'] = max_buckets
+        if suspended is not None:
+            params['suspended'] = suspended
+        if generate_key is not None:
+            params['generate-key'] = generate_key
+        if access_key is not None:
+            params['access-key'] = access_key
+        if secret_key is not None:
+            params['secret-key'] = secret_key
+        return self.proxy('PUT', 'user', params)
+
+    def set(self, uid, display_name=None, email=None, max_buckets=None,
+            suspended=None):
+        params = {'uid': uid}
+        if display_name is not None:
+            params['display-name'] = display_name
+        if email is not None:
+            params['email'] = email
+        if max_buckets is not None:
+            params['max-buckets'] = max_buckets
+        if suspended is not None:
+            params['suspended'] = suspended
+        return self.proxy('POST', 'user', params)
+
+    def delete(self, uid):
+        return self.proxy('DELETE', 'user', {'uid': uid}, json_response=False)
+
+
+@ApiController('/rgw/user/:uid/caps')
```

Comment:
Since the "caps" of a user is not a "resource" I don't think you should create a controller for this. I would add `@Resource` methods in the `RgwUser` controller for this. Example:

```python
@RESTController.Resource('POST', path='/caps')
def create_caps(self, uid, type, perm):
  # ...

@RESTController.Resource('DELETE', path="/caps")
def delete_caps(self, uid, type, perm):
  # ...
```

The above example is only possible to implement after #22528 is merged.

This comment applies to other controllers in this PR like `RgwUserKey` and `RgwUserQuota`.